sdk/core/core-util/test/public/reshape.spec.ts

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { describe, it, assert, expect } from "vitest";
+import { reshape } from "../../src/reshape";
+
+describe("reshape", function () {
+  it("renames a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c");
+    const expected = { c: 1, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", String);
+    const expected = { a: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames and maps a simple property", function () {
+    const input = { a: 1, b: 2 };
+    const actual = reshape(input, "a", "c", String);
+    const expected = { c: "1", b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renames a nested property", function () {
+    const input = { a: { c: { e: 4 }, d: 3 }, b: 2 };
+    const actual = reshape(input, "a.c.e", "f");
+    const expected = { a: { c: { f: 4 }, d: 3 }, b: 2 };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("can map elements of an array", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    const actual = reshape(input, "a.b[]", String);
+    const expected = { a: { b: ["1", "2", "3"] } };
+    assert.deepEqual(actual, expected);
+  });
+
+  it("renaming array elements is an error", function () {
+    const input = { a: { b: [1, 2, 3] } };
+    expect(() => reshape(input, "a.b[]", "oops")).toThrowError("cannot rename array indices");
+  });
+
+  it("can rename properties of elements of an array", function () {
+    const input = {
+      a: {
+        b: [
+          { c: 1, d: 2 },
+          { c: 3, d: 4 },
+        ],
+      },
+    };
+    const actual = reshape(input, "a.b[]c", "e");
+    const expected = {
+      a: {
+        b: [
+          { e: 1, d: 2 },
+          { e: 3, d: 4 },
+        ],
+      },
+    };
+    assert.deepEqual(actual, expected);
```

Comment:
I wonder if it's possible to have a case where different elements in the array needs to be reshaped differently ? like in the polymorphic base's case, it might be possible they use a base class array to represent a collection of all subtypes, but each subtypes could be different on a specific field ?

Reply:
Yeah, assuming they had different properties you could call it multiple times for each subpath:

```ts
reshape(input, "a.b[]c.d", "e");
reshape(input, "a.b[]f.g", "h");
```

One missing feature currently is the ability to select by a run-time value, like selecting based on the `kind` value of the child object. Though you could always nest the logic like

```ts
reshape(input, "a.b[]", (el) => {
   switch(el.kind) {
       case "dog": 
           return reshape(el, "birthday", Date);
           break;
       default:
           return el;
   }
});
```